### PR TITLE
Implement PartialEq and Eq for algorithm types

### PR DESCRIPTION
**Pull request by mrwhythat**
PR for #542 

`#[derive(Eq)]` seems to not work for algorithm types because of other fields (with `fn` types for example), so explicit `impl Eq for Type {}` were added.

Also, some unintended diff in `aead/chacha20_poly1305.rs` and `agreement.rs`, because I use `rustfmt` tool in my emacs. If this is a problem, I will restore it.

### CODE REVIEW

File: src/aead/aead.rs [hunk 1]
```diff
@@ -309,6 +310,19 @@ impl Algorithm {
...
+    AES_128_GCM,
+    AES_256_GCM,
+    CHACHA20_POLY1305,
+}
+
+impl PartialEq for Algorithm {
+    fn eq(&self, other: &Self) -> bool { self.id == other.id }
+}
+
+impl Eq for Algorithm {}
```
Comment (briansmith) [inline src/aead/aead.rs]: I think `Eq` can be derived and only `PartialEq` needs to be explicitly implemented, right? Assuming so, let's always derive `Eq` instead of doing `impl Eq for ...`.

File: src/aead/chacha20_poly1305.rs [hunk 2]
```diff
@@ -24,19 +24,20 @@ pub static CHACHA20_POLY1305: aead::Algorithm = aead::Algorithm {
     init: chacha20_poly1305_init,
     seal: chacha20_poly1305_seal,
     open: chacha20_poly1305_open,
+    id: aead::AlgorithmID::CHACHA20_POLY1305,
 };
 
 /// Copies |key| into |ctx_buf|.
 pub fn chacha20_poly1305_init(ctx_buf: &mut [u8], key: &[u8])
-                              -> Result<(), error::Unspecified> {
+    -> Result<(), error::Unspecified> {
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Probably we should reformat the code, but not in this PR. Please remove all the unrelated formatting changes.

File: src/agreement.rs [hunk 3]
```diff
@@ -94,8 +94,28 @@ pub struct Algorithm {
...
+
+/// Key agreement algorithm types.
+#[derive(PartialEq, Eq)]
+#[allow(non_camel_case_types)]
+pub enum AlgorithmID {
+    /// Defined in 'ec/suite_b/ecdh.rs'
+    ECDH_P256,
+    /// Defined in 'ec/suite_b/ecdh.rs'
+    ECDH_P384,
+    /// Defined in 'ec/curve25519/x25519.rs'
```
Comment (briansmith) [inline src/agreement.rs]: Let's drop all the "Defined in" comments.

File: src/agreement.rs [hunk 4]
```diff
@@ -94,8 +94,28 @@ pub struct Algorithm {
     // `ec::AgreementAlgorithmImpl` isn't public.
     #[doc(hidden)]
     pub i: ec::AgreementAlgorithmImpl,
+    #[doc(hidden)]
+    pub id: AlgorithmID,
+}
+
+/// Key agreement algorithm types.
+#[derive(PartialEq, Eq)]
+#[allow(non_camel_case_types)]
```
Comment (briansmith) [inline src/agreement.rs]: Let's put the `allow(...)` ahead of `derive(...)` in this and the other ones. No order seems better than any other, so let's things in alphabetical order so we don't have to decide another order.

File: src/ec/suite_b/ecdsa.rs [hunk 5]
```diff
@@ -50,6 +66,12 @@ enum ECDSAVerificationAlgorithmID {
     ECDSA_P384_SHA384_FIXED,
 }
 
+impl PartialEq for ECDSAVerificationAlgorithm {
+    fn eq(&self, other: &Self) -> bool { self.id == other.id }
+}
```
Comment (briansmith) [inline src/ec/suite_b/ecdsa.rs]: I think we should not implement PartialEq for this type. Instead we should implement PartialEq for `signature::VerificationAlgorithm` in some way. Maybe we need to make `signature::VerificationAlgorithm` a struct that can hold the id field first. WDYT?

File: src/aead/aead.rs (hunk 1 above)
Comment (mrwhythat) [inline src/aead/aead.rs]: Well, I tried deriving `Eq` first, but it broke the build. It seems that non-`Eq` fields make it impossible to derive. Or maybe it's the ordering. I'll research.

**Comment by mrwhythat** [comment]
@briansmith , just committed requested changes

File: src/aead/aead.rs (hunk 1 above)
Comment (mrwhythat) [inline src/aead/aead.rs]: Did some research and it seems that `derive(Eq)` simply checks for every field, it does not look for existing required impls, like Haskell does.

File: src/ec/suite_b/ecdsa.rs (hunk 5 above)
Comment (mrwhythat) [inline src/ec/suite_b/ecdsa.rs]: Well, there is `PartialEq` implementation for `ECDSASigningAlgorithm`, so implementing it for `ECDSAVerificationAlgorithm` only benefits the symmetricity. Also, `signature::VerificationAlgorithm` is a very elegant single-method protocol, which I'm a big fan of, and which modern cryptographic libraries lack, so I think coercing it to struct is a bad idea, but it may just be my irrational thinking:)

File: src/aead/aead.rs (hunk 1 above)
Comment (briansmith) [inline src/aead/aead.rs]: I looked into it. It used to work, but now it doesn't, and that is intentional: https://github.com/rust-lang/rust/issues/36830.

File: src/agreement.rs [hunk 6]
```diff
@@ -94,8 +94,25 @@ pub struct Algorithm {
...
+}
+
+/// Key agreement algorithm types.
+#[allow(non_camel_case_types)]
+#[derive(PartialEq, Eq)]
+pub enum AlgorithmID {
+    ECDH_P256,
+    ECDH_P384,
+    X25519,
 }
```
Comment (briansmith) [inline src/agreement.rs]: `CurveID` already exists, so we don't need this. I'll remove this in favor of an approach using `CurveID`.

File: src/ec/suite_b/ecdsa.rs (hunk 5 above)
Comment (briansmith) [inline src/ec/suite_b/ecdsa.rs]: > Well, there is PartialEq implementation for ECDSASigningAlgorithm

Yes, but that's because we've not yet found a solution for creating a unifying `ring::signature::SigningAlgorithm` interface.

However, we already have `ring::signature::VerificationAlgorithm`, and ultimatenly for these comparison operators to be most useful, we need to be able to compare two `VerificationAlgorithm`s, including comparing an RSA algorithm to an ECDSA algorithm. That means that `ring::signature::VerificationAlgorithm` is what ultimately needs to implement `Eq`/`PartialEq`. Let's try to find a way to do that. For now, I'll remove this part.

File: src/agreement.rs (hunk 6 above)
Comment (mrwhythat) [inline src/agreement.rs]: Removed it in favor of `algorithm.i.curve.id` comparison.

**Comment by mrwhythat** [comment]
Does it require any additional changes?

**Comment by ctz** [comment]
This will let me delete some code that computes `digest::Algorithm` equality the dumb way 👍

**Comment by briansmith** [comment]
This landed with modifications and was released in *ring* 0.12.1. The commits are:

d8134d64fa5a1109c938ebb0258b6e7aeb1b3439
eb668d5d02256ae35f13bbb80f7f0f60071126d5
b3cd49cd7f41e31faef35a29f688c20e6290d016

Thanks for submitting this and sorry for the long delay in getting it landed.

@ctz wrote:
> This will let me delete some code that computes digest::Algorithm equality the dumb way 👍

I just noticed that, and just sent you a PR: https://github.com/ctz/rustls/pull/117
